lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Copyright: © 2022 IOHK
+-- License: Apache-2.0
+--
+module Cardano.Wallet.Write.Tx.Gen
+    ( genDatum
+    , genData
+    , genDatumHash
+    , shrinkData
+    , shrinkDatum
+    , genTxOut
+    )
+    where
+
+import Prelude
+
+import Cardano.Ledger.Alonzo.Data
+    ( Data (..), dataToBinaryData )
+import Cardano.Wallet.Write.Tx
+    ( BinaryData
+    , Datum (..)
+    , DatumHash
+    , LatestLedgerEra
+    , RecentEra
+    , ShelleyLedgerEra
+    , TxOut
+    , cardanoEraFromRecentEra
+    , datumFromCardanoScriptData
+    , datumHashFromBytes
+    , datumToCardanoScriptData
+    , shelleyBasedEraFromRecentEra
+    )
+import Data.ByteString
+    ( ByteString )
+import Data.Maybe
+    ( fromMaybe )
+import Ouroboros.Consensus.Cardano.Block
+    ( EraCrypto, StandardCrypto )
+import Test.QuickCheck
+    ( Gen
+    , arbitrary
+    , choose
+    , listOf
+    , oneof
+    , resize
+    , shrinkMapBy
+    , sized
+    , vector
+    , vectorOf
+    )
+
+import qualified Cardano.Api.Gen as Cardano
+import qualified Cardano.Api.Shelley as Cardano
+import qualified Data.ByteString as BS
+import qualified Plutus.V1.Ledger.Api as PV1
+
+genDatum :: (EraCrypto era ~ StandardCrypto) => Gen (Datum era)
+genDatum = oneof
+    [ Datum <$> genData
+    , DatumHash <$> genDatumHash
+    , pure NoDatum
+    ]
+
+-- Originally from https://github.com/input-output-hk/cardano-ledger/blob/c7c63dabdb215ebdaed8b63274965966f2bf408f/eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Serialisation/Generators.hs#L66-L79
+genData :: Gen (BinaryData era)
+genData = dataToBinaryData . Data <$> resize 5 (sized gendata)
```

Comment:
@Anviking 

I'm a bit curious: why use `resize` to hard-wire the QC size parameter here?

If we specify `resize` here with a fixed constant `c`, it would presumably prevent the generator from generating values smaller than `c`, which could prevent us from discovering bugs that only occur with smaller values.

Would it be feasible to instead use `scale` here, perhaps with `div` or `mod`? Then at least we'd still be able to generate a range of different sizes (albeit a smaller range).

Alternatively, our most common pattern in other `Gen` modules seems to be:

- make `Gen` modules define generators that depend on the QC size parameter (whatever that may be)
- if we need to alter the size in some way, either:
    - use `scale` at the point where we define an `Arbitrary` instance, or
    - if we need more fine-grained control over a parameter for a specific property, either:
        - use a newtype wrapper that adjusts the size (such as `Scale{Div,Mod}`).
        - use `forAll` in combination with an ordinary size combinator (such as `scale`).

Reply:
I had just copied this from the ledger, but good point 👍 

What do you think about `scale (min 7)`? Then we allow both smaller sizes as well as slightly larger than what was previously generated.